### PR TITLE
implement attempting to reconnect with all available addresses

### PR DESCRIPTION
**Pull request by denise-opendaq**
# Brief

Pass available connection strings of the remote device to the native streaming client, enabling reconnection attempts through all available addresses. This improves reconnection reliability when a device loses one interface but remains accessible via others. 

Manually checked:
- Disabling the connection interface in both client and server
- The addressed of active clients are correct
- Primary connection was done with only ipv4, ipv6 or both

# API changes

None

# Required application changes

None

# Required module changes

None

### CODE REVIEW

**Review by NikolaiShipilov — APPROVED** [review]

**Review by JakaMohorko — COMMENTED** [review]
Upon review, I would suggest that this is reworked for native to create an internal cache of available addresses after device creation.

File: modules/native_streaming_client_module/src/native_device_impl.cpp [hunk 1]
```diff
@@ -303,6 +308,59 @@ void NativeDeviceHelper::setupProtocolClients(const ContextPtr& context)
                                               transportConnectionStatusChangedCb);
 }
 
+void NativeDeviceHelper::setupAlternativeAddressesCallback()
```
Comment (JakaMohorko) [inline modules/native_streaming_client_module/src/native_device_impl.cpp]: I think we should just cache all available addresses in `onCompleteServerCapabilities`.

**Review by JakaMohorko — CHANGES_REQUESTED** [review]

File: modules/native_streaming_client_module/src/native_device_impl.cpp [hunk 2]
```diff
@@ -466,6 +467,63 @@ void NativeDeviceImpl::removed()
...
+                    primaryAddressType = primaryAddressTypeCandidate;
+            }
+        }
+    }
+
+    DeviceInfoPtr deviceInfo;
+    OPENDAQ_RETURN_IF_FAILED(getInfo(&deviceInfo));
+
+    if (deviceInfo == nullptr)
+       return OPENDAQ_IGNORED;
```
Comment (JakaMohorko) [inline modules/native_streaming_client_module/src/native_device_impl.cpp]: I would just return the `OPENDAQ_SUCCESS` error codes, as the call we're in is actually `setComponentConfig`.

File: modules/native_streaming_client_module/src/native_device_impl.cpp [hunk 3]
```diff
@@ -466,6 +467,63 @@ void NativeDeviceImpl::removed()
...
+            continue;
+
+        if (addressInfo.getReachabilityStatus() == AddressReachabilityStatus::Unreachable)
+            continue;
+
+        alternativeAddresses.pushBack(addressInfo.getAddress());
+    }
+
+    if (!deviceHelper)
+        return OPENDAQ_IGNORED;
```
Comment (JakaMohorko) [inline modules/native_streaming_client_module/src/native_device_impl.cpp]: This can probably be checked at the top.

File: modules/native_streaming_client_module/src/native_device_impl.cpp [hunk 4]
```diff
@@ -466,6 +467,63 @@ void NativeDeviceImpl::removed()
...
+                if (primaryAddressTypeCandidate.getLength())
+                    primaryAddressType = primaryAddressTypeCandidate;
+            }
+        }
+    }
+
+    DeviceInfoPtr deviceInfo;
+    OPENDAQ_RETURN_IF_FAILED(getInfo(&deviceInfo));
+
+    if (deviceInfo == nullptr)
```
Comment (JakaMohorko) [inline modules/native_streaming_client_module/src/native_device_impl.cpp]: Needs to check for `deviceInfo.assigned()`.

File: modules/native_streaming_client_module/src/native_streaming_device_impl.cpp [hunk 5]
```diff
@@ -41,6 +41,60 @@ NativeStreamingDeviceImpl::NativeStreamingDeviceImpl(const ContextPtr& ctx,
     this->statusContainer.asPtr<IComponentStatusContainerPrivate>().addStatus("ConnectionStatus", connectionStatus);
 }
 
+ErrCode NativeStreamingDeviceImpl::setComponentConfig(IPropertyObject* config)
```
Comment (JakaMohorko) [inline modules/native_streaming_client_module/src/native_streaming_device_impl.cpp:45]: These calls are almost duplicate. Could we have them in a shared helper method? It could simply take as parameters: `config`, `deviceInfo`, `transportProtocolClient`.

**Review by JakaMohorko — DISMISSED** [review]
Looks good in general, some refactoring suggestions.

This is missing test cases, but it likely can not be tested in the GitHub environment. Should be checked manually for edge cases.

File: shared/libraries/native_streaming_protocol/src/native_streaming_client_handler.cpp [hunk 6]
```diff
@@ -276,7 +276,48 @@ void NativeStreamingClientImpl::tryReconnect()
...
+        }
+        catch (const std::exception& e)
+        {
+            LOG_W("Failed to get alternative addresses: {}", e.what());
+        }
+    }
+    else
+    {
+        LOG_D("No alternative addresses callback set for reconnection");
+    }
```
Comment (NikolaiShipilov) [inline shared/libraries/native_streaming_protocol/src/native_streaming_client_handler.cpp]: this section looks like smth from prev implementation based on callbacks.

**Review by NikolaiShipilov — COMMENTED** [review]
One major thing (I've probalby missed while doing review last time as well) I've just realized - is that there likely isn't a mechanism for providing alternative addresses to established streaming connection & created streaming object (except streaming device use-case) - addresses are available only for device object via device info. Therefore, the streaming stays disconnected even if the configuration protocol successfully switched to alternative address.

**Review by NikolaiShipilov — APPROVED** [review]
